src/script/squirrel.cpp

```diff
@@ -544,69 +547,87 @@ class SQFile {
+			this->buffer.erase(0, this->consumer.GetBytesRead());
+
+			size_t buffer_size = this->buffer.size();
+			size_t read_size = Align(min_size - buffer_size, 4096); // read pages of 4096 bytes
+			// TODO C++23: use std::string::resize_and_overwrite()
+			this->buffer.resize(buffer_size + read_size);
+			auto dest = std::span(this->buffer.data(), this->buffer.size()).subspan(buffer_size);
+			buffer_size += this->ReadInternal(dest);
+			this->buffer.resize(buffer_size);
+
+			this->consumer = StringConsumer(this->buffer);
 		}
-		if (count == 0) return 0;
-		size_t ret = fread(buf, elemsize, count, this->file);
-		this->pos += ret * elemsize;
-		return ret;
+		return this->consumer;
+	}
+
+	size_t Read(void *buf, size_t max_size)
+	{
+		std::span<char> dest(reinterpret_cast<char *>(buf), max_size);
+
+		auto view = this->consumer.Read(max_size);
+		std::copy(view.data(), view.data() + view.size(), dest.data());
+		size_t result_size = view.size();
+
+		if (result_size < max_size) {
+			assert(!this->consumer.AnyBytesLeft());
+			result_size += this->ReadInternal(dest.subspan(result_size));
+		}
+
+		return result_size;
 	}
 };
 
 static char32_t _io_file_lexfeed_ASCII(SQUserPointer file)
 {
-	unsigned char c;
-	if (((SQFile *)file)->Read(&c, sizeof(c), 1) > 0) return c;
-	return 0;
+	StringConsumer &consumer = reinterpret_cast<SQFile *>(file)->GetConsumer();
+	return consumer.TryReadUint8().value_or(0); // read as unsigned, otherwise integer promotion breaks it
 }
 
 static char32_t _io_file_lexfeed_UTF8(SQUserPointer file)
 {
-	char buffer[5];
-
-	/* Read the first character, and get the length based on UTF-8 specs. If invalid, bail out. */
-	if (((SQFile *)file)->Read(buffer, sizeof(buffer[0]), 1) != 1) return 0;
-	uint len = Utf8EncodedCharLen(buffer[0]);
-	if (len == 0) return -1;
-
-	/* Read the remaining bits. */
-	if (len > 1 && ((SQFile *)file)->Read(buffer + 1, sizeof(buffer[0]), len - 1) != len - 1) return 0;
-
-	/* Convert the character, and when definitely invalid, bail out as well. */
-	char32_t c;
-	if (Utf8Decode(&c, buffer) != len) return -1;
-
-	return c;
+	StringConsumer &consumer = reinterpret_cast<SQFile *>(file)->GetConsumer();
+	return consumer.AnyBytesLeft() ? consumer.ReadUtf8(-1) : 0;
 }
 
 static char32_t _io_file_lexfeed_UCS2_no_swap(SQUserPointer file)
 {
-	unsigned short c;
-	if (((SQFile *)file)->Read(&c, sizeof(c), 1) > 0) return (char32_t)c;
-	return 0;
+	StringConsumer &consumer = reinterpret_cast<SQFile *>(file)->GetConsumer();
+	if (!consumer.AnyBytesLeft()) return 0;
+	return FROM_LE16(consumer.ReadUint16LE(0));
 }
 
 static char32_t _io_file_lexfeed_UCS2_swap(SQUserPointer file)
 {
-	unsigned short c;
-	if (((SQFile *)file)->Read(&c, sizeof(c), 1) > 0) {
-		c = ((c >> 8) & 0x00FF)| ((c << 8) & 0xFF00);
-		return (char32_t)c;
-	}
-	return 0;
+	StringConsumer &consumer = reinterpret_cast<SQFile *>(file)->GetConsumer();
+	if (!consumer.AnyBytesLeft()) return 0;
+	return FROM_BE16(consumer.ReadUint16LE(0));
 }
```

Comment:
Not really in scope for this PR, but... the byte-order-mark used to detect these is actually for UTF-16. So UCS2 is a misnomer, and with UTF-16 you also have multi-character encodings like UTF-8 which we do not seem to care about.

Maybe check whether there are any scripts in the content system with UTF-16, and if there are none just remove UCS/UTF-8 support?

Reply:
#13992